utils/crypto.cpp

```diff
@@ -0,0 +1,236 @@
+     *
+     * @param rsa a pointer to the public key structure
+     * @param signature a const vector reference to the signature associated with the uuid
+     * @param uuid a const string reference to the nodes' uuid
+     * @param is_authentic an reference to a bool that will be set to true if the signature is valid, false otherwise.
+     * @return a bool that returns true if the validation process was successful.
+     */
+    bool
+    RSA_verify_signature(const RSA* rsa, const std::vector<unsigned char>& signature, const std::string& uuid, bool& is_authentic)
+    {
+        std::unique_ptr<EVP_PKEY, std::function<void(EVP_PKEY*)>> public_key{EVP_PKEY_new(), EVP_PKEY_free};
+        std::unique_ptr<EVP_MD_CTX, std::function<void(EVP_MD_CTX*)>> RSA_verification_context{EVP_MD_CTX_create(), EVP_MD_CTX_destroy};
+
+        // I could find no guidance in the OpenSSL documentation that suggests that the EVP_MD
+        // pointer needs to be released after use, though in the same page the documentation
+        // points out that EVP_MD_CTX pointers *do* need to be released. For now I will leave
+        // the following as a raw pointer.
+        const EVP_MD* sha256_message_digest = EVP_sha256();
+
+        is_authentic = false;
+        EVP_PKEY_assign_RSA(public_key.get(), rsa);
+
+        EVP_DigestInit_ex( RSA_verification_context.get(), sha256_message_digest, nullptr);
+
+        if (1 != EVP_DigestVerifyInit(RSA_verification_context.get(), nullptr, sha256_message_digest, nullptr, public_key.get()))
+        {
+            compose_and_log_OpenSSL_error("EVP_DigestVerifyInit");
+            return false;
+        }
+
+        if (1 != EVP_DigestVerifyUpdate(RSA_verification_context.get(), uuid.c_str(), uuid.size()))
+        {
+            compose_and_log_OpenSSL_error("EVP_DigestVerifyUpdate");
+            return false;
+        }
+        // The const cast is for the unix gcc c++ compiler
+        auto auth_status = EVP_DigestVerifyFinal(RSA_verification_context.get(), const_cast<unsigned char*>(signature.data()), signature.size());
+
+        // There are a couple of options here, the signature was successfully
+        // validated or invalidated, or something went wrong during the
+        // validation process.
+        if ( auth_status==0 || auth_status==1)
+        {
+            is_authentic = (auth_status==1);
+            return true;
+        }
+        else
+        {
+            compose_and_log_OpenSSL_error("EVP_DigestVerifyUpdate");
+            return false;
+        }
+    }
+}
+
+
+namespace bzn::utils::crypto
+{
+    std::string
+    retrieve_bluzelle_public_key_from_contract()
+    {
+        // TODO retrieve the public key from the contract
+        return temporary_public_pem;
+    }
+
+
+    int
+    base64_decode(const std::string& base64_message, std::vector<unsigned char>& decoded_message)
+    {
+        const auto decoded_length = calculate_decode_length(base64_message);
```

Comment:
this isn't the kind of places auto is used for

Reply:
Right, I'll stop using auto for pod.